File: src/main/java/com/parmet/squashlambdas/email/AppendableList.java
package com.parmet.squashlambdas.email;

import com.google.common.collect.ImmutableList;
import com.google.common.collect.Iterators;
import com.google.common.collect.UnmodifiableIterator;
import java.util.ArrayList;
import java.util.List;

class AppendableList<T> implements Appendable2<T> {
  private final List<T> delegate = new ArrayList<>();

  public AppendableList() {}

  public AppendableList(List<T> src) {
    delegate.addAll(src);
  }

  @Override
  public AppendableList<T> append(T toAppend) {
    delegate.add(toAppend);
    return this;
  }

  @Override
  public ImmutableList<T> toList() {
    return ImmutableList.copyOf(delegate);
  }

  @Override
  public UnmodifiableIterator<T> iterator() {
    return Iterators.unmodifiableIterator(delegate.iterator());
  }
}

File: src/main/kotlin/com/parmet/squashlambdas/email/AppendableList.kt
package com.parmet.squashlambdas.email

class AppendableList<T>() : Appendable<T>, Iterable<T> {
    private val delegate: MutableList<T> = mutableListOf()

    constructor(src: List<T>): this() {
        delegate.addAll(src)
    }

    override fun append(toAppend: T) = apply { delegate.add(toAppend) }

    override fun iterator() = delegate.iterator()
}
